dask/sizeof.py

```diff
@@ -135,44 +135,78 @@ def register_pandas():
     import numpy as np
     import pandas as pd
 
-    def object_size(x):
-        if not len(x):
+    from dask.dataframe._compat import dtype_eq
+
+    def object_size(*xs):
+        if not xs:
+            return 0
+        ncells = sum(len(x) for x in xs)
+        if not ncells:
             return 0
-        sample = np.random.choice(x, size=20, replace=True)
-        sample = list(map(sizeof, sample))
-        return sum(sample) / 20 * len(x)
+
+        # Deduplicate Series of references to the same objects,
+        # e.g. as produced by read_parquet
+        unique_samples = {}
+        for x in xs:
+            sample = np.random.choice(x, size=100, replace=True)
+            for i in sample.tolist():
+                unique_samples[id(i)] = i
+
+        nsamples = 100 * len(xs)
+        sample_nbytes = sum(sizeof(i) for i in unique_samples.values())
+        if len(unique_samples) / nsamples > 0.5:
+            # Less than half of the references are duplicated.
+            # Assume that, if we were to analyze twice the amount of random references,
+            # we would get twice the amount of unique objects too.
+            return int(sample_nbytes * ncells / nsamples)
+        else:
+            # Assume we've already found all unique objects and that all references that
+            # we have not yet analyzed are going to point to the same data.
+            return sample_nbytes
 
     @sizeof.register(pd.DataFrame)
     def sizeof_pandas_dataframe(df):
-        p = sizeof(df.index)
-        for _, col in df.items():
-            p += col.memory_usage(index=False)
+        p = sizeof(df.index) + sizeof(df.columns)
+        object_cols = []
+        prev_dtype = None
+
+        # Unlike df.items(), df._series will not duplicate multiple views of the same
+        # column e.g. df[["x", "x", "x"]]
+        for col in df._series.values():
+            if prev_dtype is None or not dtype_eq(prev_dtype, col.dtype):
+                prev_dtype = col.dtype
+                # Contiguous columns of the same dtype share the same overhead
+                p += 1200
+            p += col.memory_usage(index=False, deep=False)
             if col.dtype == object:
```

Comment:
Should we also handle `string[python]` here?  (totally ok to defer this to future work though, this is likely scope creep)

Reply:
I'll open a further PR for it